hal/src/gcc/core_hal.cpp

```diff
@@ -36,33 +36,61 @@
 #include "device_config.h"
 #include "hal_platform.h"
 #include <boost/crc.hpp>  // for boost::crc_32_type
+#include <sstream>
+#include <iomanip>
 
 
 using std::cout;
 
-void debug_output_(const char* msg);
+static LoggerOutputLevel log_level = NO_LOG_LEVEL;
 
 void setLoggerLevel(LoggerOutputLevel level)
 {
-    set_logger_output(debug_output_, level);
+    log_level = level;
 }
 
-extern "C" int main(int argc, char* argv[])
+void log_message_callback(const char *msg, int level, const char *category, uint32_t time, const char *file, int line,
+        const char *func, void *reserved)
 {
-    setLoggerLevel(NO_LOG_LEVEL);
-    if (read_device_config(argc, argv)) {
-        app_setup_and_loop();
+    if (level < log_level) {
+        return;
     }
-    return 0;
+    std::ostringstream strm;
```

Comment:
do we need to use stringstream here vs using sprintf()? Using sprintf() would keep the code size to a minimum.